src/exercise/04.js

```diff
@@ -65,18 +30,92 @@ function Board() {
         {renderSquare(7)}
         {renderSquare(8)}
       </div>
-      <button className="restart" onClick={restart}>
-        restart
-      </button>
     </div>
   )
 }
 
+function Info({ status, moves, currentStep, setCurrentStep }) {
+  function goToMove(step) {
+    // Note to self:
+    // When currentStep changes, all the derived states for squares/winner/etc will
+    // also change accordingly, which is why the history buttons change the board.
+    setCurrentStep(step);
+  }
+  return (
+    <div>
+      <div className="status">{status}</div>
+      <ol>
+        {moves.map((move, step) => {
+          let buttonText = step ? `Go to move #${step}` : `Go to start`;
+          let isCurrentStep = currentStep === step;
+          return (
+            <li key={move}>
+              <button
+                disabled={isCurrentStep}
+                onClick={() => goToMove(step)}
+              >
+                {buttonText} {isCurrentStep ? `(current)` : null}
+              </button>
+            </li>
+          );
+        }
+        )}
+      </ol>
+    </div>
+  );
+}
+
 function Game() {
+  const [history, setHistory] = useLocalStorageState('history', [Array(9).fill(null)]); // Set this to an empty array with nine slots.
+  const [currentStep, setCurrentStep] = useLocalStorageState('currentStep', 0);
+
+  const squares = history[currentStep]; // Use history state instead of original squares state.
+  const nextValue = calculateNextValue(squares);
+  const winner = calculateWinner(squares);
+  const status = calculateStatus(winner, squares, nextValue);
+
+  function selectSquare(square) {
+    if (winner || squares[square] ) {
+      return;
+    }
+
+    const currentSquares = [...squares];
+    currentSquares[square] = nextValue;
+
+    const moves = [...history];
+    const thisMove = moves[square] = currentSquares;
+    const newHistory = [...history, thisMove];
```

Comment:
I *think* this will break if you step backward in history and then make a move, because it will tack the next move at the end of the history array rather than taking a slice of the history array to the current index and then appending the new move.